File: src/main/java/ca/alexgirard/harpoonIJ/ShowHarpoon.java
package ca.alexgirard.harpoonIJ;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class ShowHarpoon extends AnAction {

    private static HarpoonDialog dialog;

    public static void NextHarpoonItem() {
        if (dialog == null || !dialog.isShowing())
            return;
        dialog.Next();
    }
    public static void PreviousHarpoonItem() {
        if (dialog == null || !dialog.isShowing())
            return;
        dialog.Previous();
    }
    public static void SelectHarpoonItem() {
        if (dialog == null || !dialog.isShowing())
            return;
        dialog.Ok();
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent e) {

        var stringBuilder = new StringBuilder();
        var fileStrings = HarpoonState.GetFiles(e.getProject());
        var project = e.getProject();
        var projectPath = project == null ? "" : project.getBasePath();
        projectPath = projectPath == null ? "" : projectPath;

        for (var vFile : fileStrings) {
            var path = vFile == null ? "" : vFile.getCanonicalPath();
            path = path == null ? "" : path;
            stringBuilder.append(path.replace(projectPath, "...")).append("\n");
        }
        
        var text = stringBuilder.toString().trim();
        dialog = new HarpoonDialog(text);
        var result = dialog.showAndGet();
        if (text.equals(dialog.editorTextField.getText().trim())) {
            if(result) {
                NavigateToFile(project);
            }
            return;
        }
        String newText = dialog.editorTextField.getText().trim().replace("...", projectPath);

        String[] lines = newText.split("\n");
        var outputList = new ArrayList<String>();
        for (String line : lines) {
            outputList.add(line.trim());
        }
        HarpoonState.SetFiles(outputList, e.getProject());
        if (result) {
            NavigateToFile(project);
        }
    }
    private void NavigateToFile(Project project){
        if (project == null) return;
        VirtualFile vf = HarpoonState.GetItem(dialog.SelectedIndex, project);
        if (vf == null)
            return;
        var fileManager = FileEditorManager.getInstance(project);
        fileManager.openFile(vf, true);
    }
        
}

File: src/main/kotlin/ca/alexgirard/harpoonIJ/ShowHarpoon.kt
package ca.alexgirard.harpoonIJ

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent

class ShowHarpoon : AnAction() {
    override fun actionPerformed(e: AnActionEvent) {
        val project = e.project ?: return;
        val stringBuilder = StringBuilder()
        val fileStrings = HarpoonState.files(project)
        var projectPath = project.basePath
        projectPath = projectPath ?: ""
        for (vFile in fileStrings) {
            var path = if (vFile == null) "" else vFile.canonicalPath
            path = path ?: ""
            stringBuilder.append(path.replace(projectPath, "...")).append("\n")
        }
        val text = stringBuilder.toString().trim { it <= ' ' }
        val dialog = HarpoonDialog(text)
        dialog.showAndGet()
        val newText = dialog.editorTextField!!.getText().trim { it <= ' ' }.replace("...", projectPath)
        if (text == newText) return
        val lines = newText.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val outputList = ArrayList<String?>()
        for (line in lines) {
            outputList.add(line.trim { it <= ' ' })
        }
        HarpoonState.setFiles(outputList, project)
    }
}
